Implement C++ API generator for JitBuilder

Leonardo2718 opened the pull request:
These commits implement the following

* Some JSON schema describing a format for specifying the JitBuilder API
* A JSON description of the JitBuilder API
* Some tests to validate the schema and the API description
* A generator that outputs a C++ API for JitBuilder

Although tests are included as part of this change set, they are not added to any CI builds as more work and infrastructure are needed for these to work correctly. However, I have run the tests on my local machine and all of them passed.

This work is related to pull request #2915 and issue #2397. Although #2915 was opened fist, this PR must be merged first.

Due to the nature and magnitude of this work, an Eclipse CQ is required before merging.

@mstoodle please review
@0xdaryl FYI

[ci skip] since there are no useful tests to run

mstoodle commented on jitbuilder/apigen/cppgen.py:
```diff
@@ -0,0 +1,1066 @@
+    client object. Any other parameters accepting an object
+    type also use opaque pointers.
+
+    The generated list is usable as the parameter delcaraion
```
typo: delcaraion

Leonardo2718 commented on jitbuilder/apigen/jitbuilder.api.json:
```diff
@@ -28,14 +28,1101 @@
+                , "parms": [ {"name":"b","type":"IlBuilder"} ]
+                },
+                { "name": "Goto"
+                , "overloadsuffix": "New"
```
Off the top of my head, maybe "Builder" or "NewBuilder"?

mstoodle replied:
Let's keep thinking :)

Leonardo2718 commented:
> It seemed like there are parts of cppgen.py that aren't really "cpp" specific but more like utilities that make it easier to manipulate/identify features in the API definition. Maybe some of that code could be moved into a separate python file to ease the creation of other language bindings?

Yes, my plan was to do this soon after this PR was merged as part of a larger refactor. But, I guess now is as good a time as any :slightly_smiling_face: .

> I notice you aren't including any documentation in the initial API definitions. Perhaps we should standardize the Doxygen formatting and define some kind of meta characters that can be used to properly format the doc as comments in the generated API "header" files ? While not every service has documentation right now, I think there are some that have reasonably good doc and I would prefer to capture it here and have it included in the generated files that people will, ultimately, program against. Alternatively, the generator could collect all the doc together into a more readable form than json. I'm open to ideas; main goal is to be able to give people an API doc.

Agreed. I haven't included any documentation as I'm unsure about how to do it elegantly. A big part of the problem is that JSON does not allow multi-line strings. A common work-around is to use a list of single-line strings instead. I'm a bit concerned about the level of noise this would create though. An alternative could be to store the documentation in a separate file encoded in a "friendlier" format such as YAML. Checking that the documents are in sync (in terms of what exists, what things are called, etc) should be straightforward to automate. However, I would also prefer for all the information related to an API to be in one place. (As a side note, I considered using YAML to specify the API but chose JSON instead because the latter seems to have better / more consistent support across languages). I am open to other people's thoughts on this. @charliegracie, since you've already volunteered to work on Ruby and Pythin bindings using this, do you have any preferences?

Leonardo2718 commented:
I've created a branch where I added documentation to a few fields using lists of strings. Here's what it looks like:

https://github.com/Leonardo2718/omr/blob/dc6da2f79073cdbde148ac125d6bd110ae7db9fb/jitbuilder/apigen/jitbuilder.api.json#L1798-L1871

Thoughts?

Leonardo2718 commented:
Thanks @mstoodle.

I've run the code through a spell checker and fixed every mistake it reported. Hopefully that got most of them.

I've also run all the tests I've implemented so far (but haven't pushed yet) and all the JitBuilder tests and examples (which depend on #2915 ) and everything passes.

I will be adding API documentation in a later PR, following the style I used in the link above.

I think this is ready to be merged!
